making-mods/advanced-js/mutation-observers.md

```diff
@@ -0,0 +1,199 @@
+
+  - Sometimes elements are not hidden or transformed into something else when a change to the UI is made. In these cases, new elements are added, so an observer of this type is useful to detect those changes.
+  - Adding `subtree: true` as well, like so `{childList: true, subtree: true}`, will extend the observer to all decedent elements of the `targetElement`rather than only the first layer of children.
+    - So all the elements below the `.parent` will be observed with this configuration (denoted by the class `observed`):
+
+      ```HTML
+      <div class="parent targetElement">
+        <div class="child observed">
+          <div class="grandChild observed"> ... </div>
+        </div>
+        <div class="child observed"> ... </div>
+      </div>
+
+      <div class="notObserved"> ... </div>
+      ```
+
+### Handler function for mutations (`callBack()`)
+
+The handler function is run whenever a mutation of the types you configured is detected.
+
+The list of mutations should be passed in as a parameter to allow the function to perform different actions depending on the contents of the mutations. In the example, this parameter is named `mutations` and included like so:
+
+```JavaScript
+function callBack (mutations) {
+  ...
+```
+
+There is often more than one mutation included in the list, so it is important to check each one individually. This is accomplished with a `forEach` loop in the example (usage information can be found [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach)).
+
+Once an individual mutation is isolated, you can check what type of mutation it is by looking at the `type` attribute of the mutation object ( `if (mutation.type === ... ) { ...` ). The types that are possible are determined by the configuration of the observer. The `type` could be equal to `attribute` or `childList` based on the configuration options outlined above.
+
+When it is determined that the mutation is the correct type that you are looking for, you can then do any checks on it that you wish and run any actions accordingly. See the code examples below to see various implementations that describe what you can do. The [fullscreen observer](#fullscreen-observer) shows an example of acting on an attribute change and the [address bar / mail bar observer](#address-bar--mail-bar-observer) shows checking `addedNodes` with a childList observer.
+
+## Code examples
+
+The following are some examples of Mutation Observers that could be useful in the creation of mods. They start with the full code and then include an explanation of how they work.
+
+### Fullscreen observer
+
+// TODO: write intro
+
+// TODO: generalize and comment code
+
+```JavaScript
+let browser = document.getElementById("browser");
+let oldState =
+  browser.classList.contains("fullscreen") ||
+  browser.classList.contains("minimal-ui");
+let fullscreenObserver = new MutationObserver(function (mutations) {
```

Comment:
You don’t follow your ‘better practice’ here 😉 (but IMO this way it’s better).

Reply:
It really is "better" to define everything separately for readability, but I too prefer just defining it in place. But if I am trying to teach people, I should at least try to steer them in that direction. Do as I say, not as I do 😁

I will probably define the handler functions when I get around to finishing the examples section.